[Feature] support report for large dataset

xiaoyong-z opened the pull request:
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #9908

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 23 / 117 (19.66%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[802]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|0|2|00.00%|[114, 583]|
|:large_blue_circle:|com/starrocks/planner/FileScanNode.java|0|2|00.00%|[670, 711]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/execution/LoadExecutor.java|0|1|00.00%|[26]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|4|33|12.12%|[281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 309, 310, 311, 312, 313, 355, 356]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|8|12.50%|[362, 363, 364, 365, 366, 367, 369]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|12|28|42.86%|[347, 354, 358, 359, 360, 361, 362, 367, 368, 369, 370, 372, 373, 376, 377, 379]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|2|4|50.00%|[129, 131]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/BulkLoadJob.java|1|1|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 76 / 163 (46.63%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[802]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|8|13|61.54%|[164, 186, 598, 601, 602]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|22|30|73.33%|[126, 128, 129, 130, 131, 132, 136, 138]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|18|20|90.00%|[1367, 1368]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 76 / 163 (46.63%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[802]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|8|13|61.54%|[164, 186, 598, 601, 602]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|22|30|73.33%|[126, 128, 129, 130, 131, 132, 136, 138]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|18|20|90.00%|[1367, 1368]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

meegoo commented on be/src/common/config.h:
```diff
@@ -220,6 +220,9 @@ CONF_mDouble(memory_ratio_for_sorting_schema_change, "0.8");
 CONF_mInt32(disk_stat_monitor_interval, "5");
+CONF_mInt32(profile_report_interval, "10");
+// load_report_interval should be an integer multiple of profile_report_interval
+CONF_mInt32(load_report_interval, "30");
```
suggest calculate by profile_report_interval, minimize user config

meegoo commented on be/src/exec/pipeline/fragment_context.cpp:
```diff
@@ -41,14 +41,20 @@ FragmentContext* FragmentContextManager::get_or_register(const TUniqueId& fragme
 
     if (_fragment_contexts.find(fragment_id) != _fragment_contexts.end()) {
-        return;
+        std::stringstream msg;
```
Behavior is not the same as before, is there any problem?

xiaoyong-z replied:
In previous code, the FragmentExecutor will continue to execute if the _fragment_contexts have been registered, this will lead to the same fragment  repeatedly executed for two times.
If the fragment context has been registerd before, we should abort the FragmentExecutor, the following prepare function should fail.
```
Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParams& common_request,
                                 const TExecPlanFragmentParams& unique_request) {
    DCHECK(common_request.__isset.desc_tbl);
    DCHECK(common_request.__isset.fragment);

    UnifiedExecPlanFragmentParams request(common_request, unique_request);

    bool prepare_success = false;
    DeferOp defer([this, &prepare_success]() {
        if (!prepare_success) {
            _fail_cleanup();
        }
    });
    RETURN_IF_ERROR(exec_env->query_pool_mem_tracker()->check_mem_limit("Start execute plan fragment."));

    RETURN_IF_ERROR(_prepare_query_ctx(exec_env, request));
    RETURN_IF_ERROR(_prepare_fragment_ctx(request));
    RETURN_IF_ERROR(_prepare_workgroup(request));
    RETURN_IF_ERROR(_prepare_runtime_state(exec_env, request));
    RETURN_IF_ERROR(_prepare_exec_plan(exec_env, request));
    RETURN_IF_ERROR(_prepare_global_dict(request));
    RETURN_IF_ERROR(_prepare_pipeline_driver(exec_env, request));

    RETURN_IF_ERROR(_query_ctx->fragment_mgr()->register_ctx(request.fragment_instance_id(), _fragment_ctx));
    prepare_success = true;

    return Status::OK();
}
```

meegoo commented on be/src/storage/storage_engine.cpp:
```diff
@@ -330,6 +332,50 @@ void StorageEngine::_start_disk_stat_monitor() {
     }
 }
 
+void StorageEngine::_start_report_profile() {
```
It is not very suitable as a function of storage engine, report is for long-time/batch processing.

xiaoyong-z committed:
[Feature] support report for large dataset

1. add progress report for long run broker load and insert into in pipeline and nonpipeline model
2. broker load progress = read_bytes (file scan node) / file_bytes * 100
3. insert into select progress = read_row (from all olap scan node) / estimate_rows for all olap scan node(may be inaccurate)
4. insert into values progress = read_row (olap sink node) / values cardinal

Signed-off-by: xyz <a997647204@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 82 / 177 (46.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[812]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|8|12|66.67%|[164, 186, 601, 602]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|24|35|68.57%|[1293, 1315, 1388, 1389, 1395, 1396, 1397, 1398, 1400, 1401, 1402]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|22|30|73.33%|[126, 128, 129, 130, 131, 132, 136, 138]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 82 / 177 (46.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[812]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|8|12|66.67%|[164, 186, 601, 602]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|24|35|68.57%|[1293, 1315, 1388, 1389, 1395, 1396, 1397, 1398, 1400, 1401, 1402]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|22|30|73.33%|[126, 128, 129, 130, 131, 132, 136, 138]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

chaoyli commented:
run starrocks_admit_test

xiaoyong-z commented:
run starrocks_admit_test

wanpengfei-git commented:
run starrocks_admit_test

xiaoyong-z commented:
run starrocks_fe_unittest

xiaoyong-z commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 85 / 182 (46.70%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[812]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|10|16|62.50%|[159, 162, 170, 192, 607, 608]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|24|35|68.57%|[1293, 1315, 1388, 1389, 1395, 1396, 1397, 1398, 1400, 1401, 1402]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|23|31|74.19%|[130, 132, 133, 134, 135, 136, 140, 142]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

xiaoyong-z commented:
run starrocks_admit_test

xiaoyong-z commented:
@mergify rebase

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 85 / 182 (46.70%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|33|00.00%|[143, 144, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 179]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[812]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|0|1|00.00%|[224]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|10|10.00%|[363, 365, 366, 367, 368, 369, 370, 374, 376]|
|:large_blue_circle:|com/starrocks/load/EtlStatus.java|20|47|42.55%|[350, 357, 361, 362, 364, 365, 366, 367, 368, 372, 373, 374, 375, 376, 384, 389, 390, 391, 392, 394, 395, 398, 399, 401, 427, 453, 456]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|1|2|50.00%|[131]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|10|16|62.50%|[159, 162, 170, 192, 607, 608]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|24|35|68.57%|[1293, 1315, 1388, 1389, 1395, 1396, 1397, 1398, 1400, 1401, 1402]|
|:large_blue_circle:|com/starrocks/load/loadv2/InsertLoadJob.java|23|31|74.19%|[130, 132, 133, 134, 135, 136, 140, 142]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|5|5|100.00%|[]|

xiaoyong-z commented:
run starrocks_admit_test

xiaoyong-z commented:
run starrocks_admit_test

chaoyli commented:
@merfigy backport branch-2.4
